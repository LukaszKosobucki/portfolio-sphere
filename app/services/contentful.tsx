import { createClient } from "contentful";

export const fetchSpecificEntries = async ({
  content_type,
  title,
  multiple = false,
  order,
}: {
  content_type: string;
  title?: string;
  multiple?: boolean;
  order?: string;
}) => {
  const client = createClient({
    space: process.env.CONTENTFUL_SPACE_ID ?? "",
    environment: "master", // defaults to 'master' if not set
    accessToken: process.env.CONTENTFUL_ACCESS_TOKEN ?? "",
  });

  const query = {
    content_type,
    ...(title && { "fields.title": title }),
    limit: multiple ? 100 : 1,
    order,
  };

  const entries = await client.getEntries(query).then((res) => res.items);

  return entries;
};
