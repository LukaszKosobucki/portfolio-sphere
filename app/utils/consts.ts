export type Socials = {
  NAME: string;
  ICON: string;
  TEXT: string;
  HREF: string;
}[];
export type Page = {
  TITLE: string;
  DESCRIPTION: string;
};

export interface Site extends Page {
  AUTHOR: string;
}

export const WORK: Page = {
  TITLE: "Work",
  DESCRIPTION: "Places I have worked.",
};

// Projects Page
export const PROJECTS: Page = {
  TITLE: "Projects",
  DESCRIPTION: "Recent projects I have worked on.",
};
