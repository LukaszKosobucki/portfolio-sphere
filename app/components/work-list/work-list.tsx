import { fetchSpecificEntries } from "@/app/services/contentful";
import { WORK } from "@/app/utils/consts";
import Container from "../navbar/container";
import WorkItem from "./work-item/work-item";

async function WorkList() {
  const works = await fetchSpecificEntries({
    content_type: "experience",
    multiple: true,
    order: "-fields.dateStart",
  });
  return (
    <main id="work" className="scroll-mt-16">
      <div className="pt-5 pb-5">
        <Container size="md">
          <div className="animate page-heading">{WORK.TITLE}</div>
        </Container>
      </div>

      <div className="flex-1 py-5">
        <Container size="md">
          <ul>
            {works.map((work) => (
              <WorkItem key={work.fields.name as string} entry={work} />
            ))}
          </ul>
        </Container>
      </div>
    </main>
  );
}

export default WorkList;
