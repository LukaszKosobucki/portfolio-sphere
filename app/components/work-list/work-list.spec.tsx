import { fetchSpecificEntries } from "@/app/services/contentful";
import { WORK } from "@/app/utils/consts";
import { formatSnapshot } from "@/app/utils/tests/format-snapshot";
import { mockWorkList } from "@/app/utils/tests/mocks/work-list.mock";
import { documentToReactComponents } from "@contentful/rich-text-react-renderer";
import { render, waitFor } from "@testing-library/react";
import { afterAll, describe, expect, jest, mock, test } from "bun:test";
import WorkList from "./work-list";

mock.module("@/app/services/contentful", () => ({
  fetchSpecificEntries: jest.fn(),
}));

mock.module("@contentful/rich-text-react-renderer", () => ({
  documentToReactComponents: jest.fn(),
}));

describe("WorkList Component", () => {
  afterAll(() => {
    jest.clearAllMocks();
  });

  test("should match snapshot", async () => {
    (fetchSpecificEntries as jest.Mock).mockResolvedValue(mockWorkList);
    (documentToReactComponents as jest.Mock).mockReturnValue(<p>Mocked rich text content</p>);

    const { asFragment, getByText } = render(await WorkList());

    await waitFor(() => {
      expect(getByText("YouGov")).toBeInTheDocument();
    });

    expect(formatSnapshot(asFragment())).toMatchSnapshot();
  });

  test("should render a list of work items", async () => {
    (fetchSpecificEntries as jest.Mock).mockResolvedValue(mockWorkList);
    const { getByText } = render(await WorkList());

    expect(getByText(WORK.TITLE)).toBeInTheDocument();

    await waitFor(() => {
      expect(getByText("YouGov")).toBeInTheDocument();
      expect(getByText("Capgemini")).toBeInTheDocument();
    });

    expect(fetchSpecificEntries).toHaveBeenCalledWith({
      content_type: "experience",
      multiple: true,
      order: "-fields.dateStart",
    });
  });

  test("should handle empty work list", async () => {
    (fetchSpecificEntries as jest.Mock).mockResolvedValue([]);

    const { queryByText, getByText } = render(await WorkList());

    await waitFor(() => {
      expect(queryByText("YouGov")).not.toBeInTheDocument();
      expect(queryByText("Capgemini")).not.toBeInTheDocument();
    });

    expect(getByText(WORK.TITLE)).toBeInTheDocument();
  });
});
