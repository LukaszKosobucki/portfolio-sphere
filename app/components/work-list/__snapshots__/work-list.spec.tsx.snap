// Bun Snapshot v1, https://goo.gl/fbAQLP

exports[`WorkList Component should match snapshot 1`] = `
"<DocumentFragment>
  <main
    class="scroll-mt-16"
    id="work"
  >
    <div
      class="pt-5 pb-5"
    >
      <div
        class="w-full h-full mx-auto px-5 max-w-screen-md"
      >
        <div
          class="animate page-heading"
        >
          Work
        </div>
      </div>
    </div>
    <div
      class="flex-1 py-5"
    >
      <div
        class="w-full h-full mx-auto px-5 max-w-screen-md"
      >
        <ul>
          <li
            class="animate border-b border-black/10 dark:border-white/25 mt-4 first-of-type:mt-0 first-of-type:pt-0 last-of-type:border-none"
          >
            <div
              class="text-sm uppercase mb-4"
            >
              Sep 2023 - present
            </div>
            <div
              class="text-black dark:text-white font-semibold"
            >
              YouGov
            </div>
            <div
              class="text-sm font-semibold"
            >
              Junior Software Engineer
            </div>
            <article
              class="prose dark:prose-invert prose-p:text-justify"
            >
              <p>
                Mocked rich text content
              </p>
            </article>
          </li>
          <li
            class="animate border-b border-black/10 dark:border-white/25 mt-4 first-of-type:mt-0 first-of-type:pt-0 last-of-type:border-none"
          >
            <div
              class="text-sm uppercase mb-4"
            >
              Sep 2022 - Aug 2023
            </div>
            <div
              class="text-black dark:text-white font-semibold"
            >
              Capgemini
            </div>
            <div
              class="text-sm font-semibold"
            >
              Intern Software Engineer
            </div>
            <article
              class="prose dark:prose-invert prose-p:text-justify"
            >
              <p>
                Mocked rich text content
              </p>
            </article>
          </li>
        </ul>
      </div>
    </div>
  </main>
</DocumentFragment>"
`;
